src/coreclr/jit/rationalize.cpp

```diff
@@ -398,23 +398,20 @@ void Rationalizer::RewriteAssignment(LIR::Use& use)
 #ifdef FEATURE_SIMD
         if (varTypeIsSIMD(location) && assignment->OperIsInitBlkOp())
         {
-            if (location->OperGet() == GT_LCL_VAR)
+            if (location->OperIs(GT_LCL_VAR))
             {
-                var_types   simdType        = location->TypeGet();
-                GenTree*    initVal         = assignment->AsOp()->gtOp2;
-                CorInfoType simdBaseJitType = comp->getBaseJitTypeOfSIMDLocal(location);
-                if (simdBaseJitType != CORINFO_TYPE_UNDEF)
```

Comment:
we have never had `simdBaseJitType == CORINFO_TYPE_UNDEF` on x64, otherwise it would be an assert later on 
```
codegenxarch.cpp:4439:
assert(varTypeUsesFloatReg(targetType) == varTypeUsesFloatReg(op1Type));
```
from a tree like:
```
STORE_LCL_VAR simd16 V01
\--* CNS_INT 0
```

Reply:
This is possible, but we handle this in register allocation by lying about the type:
https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/lsra.cpp#L6995-L7000

```cpp
    if (simdNode->GetSimdBaseJitType() == CORINFO_TYPE_UNDEF)
    {
        // There are a few scenarios where we can get a LCL_VAR which
        // doesn't know the underlying baseType. In that scenario, we
        // will just lie and say it is a float. Codegen doesn't actually
        // care what the type is but this avoids an assert that would
        // otherwise be fired from the more general checks that happen.
        simdNode->SetSimdBaseJitType(CORINFO_TYPE_FLOAT);
    }
```

Namely there are a variety of scenarios where the base type might not get copied around for things like SIMD locals and where we lose the class handle and type information.

Ideally, we would preserve this 100% of the time, since its needed by things like CSE.

Reply:
The potential issue was:
ASG SIMD(LCL_VAR , 0) with `simdBaseJitType == CORINFO_TYPE_UNDEF` comes to rationalize and does not create `STORE_LCL_VAR(SIMD(0))` leaving `STORE_LCL_VAR(0)`.
code in lsra does not matter, it won't create a new node because it is after rationalize.
codegenxarch fails.

These lyings could be also deleted now (will push a change soon), maybe it will require to add `if ((simdNode->gtSIMDIntrinsicID == SIMDIntrinsicUpperSave) || (simdNode->gtSIMDIntrinsicID != SIMDIntrinsicUpperRestore))` to `simdcodegenxarch.cpp:genSIMDIntrinsic ` as we have this condition for arm64.

Reply:
> `ASG SIMD(LCL_VAR , 0)`

Is an interesting tree. In my testing (wrote a checker that looks for all `ASG SIMD` trees with mismatched types of LHS and RHS and ran it through SPMI), it only comes up when you `initobj/initblk` a SIMD local, and even then morph transforms it into `ASG SIMD(LCL_VAR, SIMD(0))`. I am curious to know if you know of other sources because my (WIP) fix for #51500 is to never import such `ASG`s in the first place.

Reply:
oh, this code exists here to catch what morph did not for some reason, for example:

[repro-7984.zip](https://github.com/dotnet/runtime/files/6430413/repro-7984.zip)
unzip it and run like
```
"D:\Sergey\git\runtime\artifacts\tests\coreclr\windows.x64.Checked\Tests\Core_Root\superpmi.exe" D:\Sergey\git\runtime\artifacts\obj\coreclr\windows.x64.Checked\ide\jit\Debug\clrjit.dll(or any jit that you want to use) "repro-7984.mc" -jitoption force AltJit= -jitoption force AltJitNgen= -jitoption force JitDump=*   -jitoption force NgenDump=* 
```
to see the dump, it will show you that in this case the issue is in fgMorphBlock that does not call morphTree for the trees that it creates, @kunalspathak recently pointed me to another issue caused by the same source. 
I have decided not to hack `fgMorphBlock ` anymore to squeeze positive diffs(and support struct enreg), so I am rewriting it now. However, it will take time and for this PR I want to keep this "catch morph failures" block in place.

Reply:
> I have decided not to hack `fgMorphBlock` anymore to squeeze positive diffs(and support struct enreg), so I am rewriting it now. However, it will take time.

I believe you 😄. Good luck!

>  it will show you that in this case the issue is in fgMorphBlock that does not call morphTree for the trees that it creates

Thanks!